crates/matrix-sdk-crypto/src/backups/mod.rs

```diff
@@ -477,25 +467,20 @@ impl BackupMachine {
         request_id: &TransactionId,
     ) -> Result<(), CryptoStoreError> {
         let mut request = self.pending_backup.write().await;
-
         if let Some(r) = &*request {
             if r.request_id == request_id {
-                let sessions: Vec<_> = self
-                    .store
-                    .get_inbound_group_sessions()
-                    .await?
-                    .into_iter()
-                    .filter(|s| r.session_was_part_of_the_backup(s))
+                let room_and_session_ids: Vec<(&RoomId, &str)> = r
+                    .sessions
+                    .iter()
+                    .flat_map(|(room_id, sender_key_to_session_ids)| {
+                        std::iter::repeat(room_id).zip(sender_key_to_session_ids.values().flatten())
+                    })
+                    .map(|(room_id, session_id)| (room_id.as_ref(), session_id.as_str()))
                     .collect();
 
-                for session in &sessions {
-                    session.mark_as_backed_up();
-                }
-
                 trace!(request_id = ?r.request_id, keys = ?r.sessions, "Marking room keys as backed up");
 
-                let changes = Changes { inbound_group_sessions: sessions, ..Default::default() };
-                self.store.save_changes(changes).await?;
+                self.store.mark_inbound_group_sessions_as_backed_up(&room_and_session_ids).await?;
```

Comment:
Have you considered adding a new field to `Changes`? that could be the vector of (roomId,sessionId) tupple? something like `sessions_marked_as_backed_up`? 
I think there is a sort of lock mecanism in save_changes, would be better to keep it no?

Reply:
I was following the pattern of `reset_backup_state`. We do use a transaction in sqlite and indexeddb - what kind of lock mechanism are you talking about?

I understood `save_changes` to be for writing complete objects to the store, not updating properties of existing objects, but I might be wrong. Maybe @bnjbvr (or someone else) can say what the general interface style of the crypto store is?

Reply:
Reading a bit more in https://github.com/matrix-org/matrix-rust-sdk/issues/2000#issuecomment-1572213914 - it seems to me that the pseudo-transaction mechanism of `SqliteCryptoStore::save_changes_lock` (which I think is what you are talking about) was mostly to support SLED and is planned to be removed in favour of exposing the underlying db's transations to the user.

So I don't think we need to go via this mechanism here, but I could be wrong.

Reply:
Ok, if reset_backup_state is doing it, let's do the same